src/modm/driver/adc/ads816x_impl.hpp

```diff
@@ -0,0 +1,155 @@
+// coding: utf-8
+// ----------------------------------------------------------------------------
+/*
+ * Copyright (c) 2021, Raphael Lehmann
+ *
+ * This file is part of the modm project.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+// ----------------------------------------------------------------------------
+
+#ifndef MODM_ADS816X_HPP
+#	error "Don't include this file directly! Use 'ads816x.hpp' instead."
+#endif
+
+//#include <bit>
+//#include <span>
+
+namespace modm
+{
+
+template <typename SpiMaster, typename Cs>
+ResumableResult<void>
+Ads816x<SpiMaster, Cs>::initialize(Mode mode)
+{
+	using namespace std::chrono_literals;
+	RF_BEGIN();
+
+	// Unlock register access by writing magic byte
+	RF_CALL(registerAccess(Command::Write, Register::REG_ACCESS, 0b1010'1010));
+
+	// Power up everything except the internal REFby2 buffer
+	RF_CALL(registerAccess(Command::Write, Register::PD_CNTL, 0b0001'0000));
+
+	// Data type: Only ADC value
+	RF_CALL(registerAccess(Command::Write, Register::DATA_CNTL, uint8_t(DataFormat::OnlyResult)));
+
+	// V_ref = 4.096V (internal reference)
+	RF_CALL(registerAccess(Command::Write, Register::OFST_CAL, 0b010));
+
+	// AIN0 ... AIN7 are separate channels
+	RF_CALL(registerAccess(Command::Write, Register::AIN_CFG, 0x00));
+
+	// All individual channels are single-ended inputs; connect the AIN-COM pin to GND
+	RF_CALL(registerAccess(Command::Write, Register::COM_CFG, 0x00));
+
+	// Set mode
+	RF_CALL(registerAccess(Command::Write, Register::DEVICE_CFG, uint8_t(mode)));
+
+	RF_END();
+}
+
+template <typename SpiMaster, typename Cs>
+ResumableResult<uint16_t>
+Ads816x<SpiMaster, Cs>::manualModeConversion(uint8_t afterNextChannel)
+{
+	RF_BEGIN();
+
+	RF_WAIT_UNTIL(this->acquireMaster());
+	Cs::reset();
+
+	buffer[0] = uint8_t(Command::Write) | (uint16_t(Register::CHANNEL_ID) >> 8);
+	buffer[1] = uint16_t(Register::CHANNEL_ID);
+	buffer[2] = (afterNextChannel & 0b0000'0111);
+
+	RF_CALL(SpiMaster::transfer(buffer, buffer2, 3));
+
+	if (this->releaseMaster()) {
+		Cs::set();
+	}
+
+	RF_END_RETURN( static_cast<uint16_t>((static_cast<uint16_t>(buffer2[0]) << 8) | buffer2[1]) );
+}
+
+/*
+// FIXME: Does not work correctly...
+template <typename SpiMaster, typename Cs>
+template<std::size_t N>
+ResumableResult<void>
+Ads816x<SpiMaster, Cs>::autoSequenceConversion(uint8_t channelsBitmask, std::span<uint16_t, N> result)
+{
+	using namespace std::chrono_literals;
+	RF_BEGIN();
+
+	if (std::popcount(channelsBitmask) > N) {
```

Comment:
Just a side-note, if you are ever implementing the auto conversion mode: The most sensible way to use span here would be a `std::span<uint16_t>` with a dynamic size and then have `result.size()` here instead of `N`. That way it would be a normal runtime parameter for the size. The check is at runtime anyway.